### PR TITLE
Re-write pickers to allow weighted selection

### PR DESCRIPTION
**Pull request by djkaty**
Add weighted_card_choice() to cards/utils.py
Add weighted pick test to tests/test_dsl.py

Unnecessary fluff removed.

Stolen Winter Veil's Gift is (temporarily) broken.

### CODE REVIEW

File: fireplace/cards/utils.py [hunk 1]
```diff
@@ -38,6 +38,50 @@
...
+		for i, c in enumerate(card_sets):
+			weightedsegmentstarts[i] = len(cardlist)
+			cardlist += c
+			# higher weighted sets (segments) occupy more space per item on the weight line
+			weightlinelength += len(c) * weights[i]
+
+		ret = []
+
+		# for each card
+		for c in range(count):
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: This portion can be considerably simplified. A good example is here: http://stackoverflow.com/a/4322940

Start with a list of (card, weight) pairs and just pass to the weighted_choices function.

File: fireplace/cards/utils.py [hunk 2]
```diff
@@ -38,6 +38,50 @@
...
+		# (user does not have to normalize weighting from 0-1)
+		cardlist = []
+		weightedsegmentstarts = [0] * len(card_sets)
+		weightlinelength = 0
+
+		for i, c in enumerate(card_sets):
+			weightedsegmentstarts[i] = len(cardlist)
+			cardlist += c
+			# higher weighted sets (segments) occupy more space per item on the weight line
+			weightlinelength += len(c) * weights[i]
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: This is assigned inside the for loop, but it only seems to be accessed after the loop ?

File: fireplace/cards/utils.py [hunk 3]
```diff
@@ -38,6 +38,50 @@
 AT_MAX_MANA = lambda s: MANA(s) == 10
 
 
+# take a list of weights and a list of card pools and produce a random weighted sample without replacement
+def weighted_card_choice(source, weights, card_sets, count):
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: Can we please document the types of the input args (and their sizes, etc. in cases where they would need to match one another), and optionally, use type annotations where appropriate?

File: fireplace/cards/utils.py [hunk 4]
```diff
@@ -38,6 +38,50 @@
...
+			# higher weighted sets (segments) occupy more space per item on the weight line
+			weightlinelength += len(c) * weights[i]
+
+		ret = []
+
+		# for each card
+		for c in range(count):
+
+			# pick a point along the weight line
+			r = random.random() * weightlinelength
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: Use `random.uniform` (https://docs.python.org/2/library/random.html#random.uniform)

File: fireplace/cards/utils.py [hunk 5]
```diff
@@ -38,6 +38,50 @@
...
+			pos = 0
+
+			# figure out which segment it's in
+			for i, w in enumerate(weights):
+				segmentarraylength = ((len(cardlist) if i == len(weights)-1 else weightedsegmentstarts[i+1]) - weightedsegmentstarts[i])
+				segmentlinelength = segmentarraylength * w
+				# it's in this segment, find exact card
+				if (r < pos + segmentlinelength):
+					# find percentage along this segment we're at
+					spc = (r - pos) / segmentlinelength
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: It's not clear to me whether you're intending to do floored integral division, or whether you really want floats (here and other places). Could you clarify?

File: fireplace/cards/utils.py (hunk 3 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: Done

File: fireplace/cards/utils.py (hunk 2 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: I'm lost. Everything accessed in the for loop is assigned before-hand (weightedsegmentstarts, weightlinelength and cardlist)

File: fireplace/cards/utils.py (hunk 4 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: That would give a small chance of r == weightlinelength which would be out of bounds. random.random() will never return 1.0.

File: fireplace/cards/utils.py (hunk 5 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: We really want floats. I would have used // for floored integer division.

**Comment by djkaty** [comment]
Simplified weighted choice algorithm implemented as discussed with JimboHS in IRC

File: fireplace/cards/utils.py [hunk 6]
```diff
@@ -39,52 +41,40 @@
...
+	for i in range(count):
+		# choose a set according to weighting
+		chosen_set = bisect(cum_weights, random.random() * totalweight)
+
+		# choose a random card from that set
+		chosen_card_index = random.randint(0, len(card_sets[chosen_set])-1)
+
+		chosen_cards.append(card_sets[chosen_set].pop(chosen_card_index))
+		totalweight -= weights[chosen_set]
+		cum_weights[chosen_set:] = [x - weights[chosen_set] for x in cum_weights[chosen_set:]]
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: This can probably be written as `cum_weights[chosen_set:] = cum_weights[chosen_set:] - weights[chosen_set]`

Comment (djkaty) [inline fireplace/cards/utils.py]: Aha, wasn't aware of that syntax

File: fireplace/cards/utils.py [hunk 7]
```diff
@@ -39,52 +41,40 @@
...
-				weightedsegmentstarts[i+1:] = [x-1 for x in weightedsegmentstarts[i+1:]]
-
-				break
-
-			pos += segmentlinelength
-
-	return [source.controller.card(card, source=source) for card in ret]
+	for i in range(count):
+		# choose a set according to weighting
+		chosen_set = bisect(cum_weights, random.random() * totalweight)
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: I'm pretty sure `random.random() * totalweight` just simplifies to `random.randint(0, totalweight - 1)`. I might be off by one there -- depends on the edge behavior of bisect.

File: fireplace/cards/utils.py (hunk 6 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: Are you sure?

>>> s = [1,2,3,4]
>>> s[1:] -= 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for -=: 'list' and 'int'
>>> s[1:] = s[1:] - 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for -: 'list' and 'int'
>>> s[1:] = [x - 1 for x in s]
>>> s
[1, 0, 1, 2, 3]
>>>

Comment (djkaty) [inline fireplace/cards/utils.py]: Or even

>>> s[1:] = [x-1 for x in s[1:]]
>>> s
[1, 1, 2, 3]

heh

File: fireplace/cards/utils.py [hunk 8]
```diff
@@ -39,52 +41,40 @@
...
 	if len(card_sets)==1:
 		return [source.controller.card(card, source=source) for card in random.sample(card_sets[0], count)]
 
-	# get all the cards matching each set of filters and create the weight line
-	# (user does not have to normalize weighting from 0-1)
-	cardlist = []
-	weightedsegmentstarts = [0] * len(card_sets)
-	weightlinelength = 0
+	# special case for all weights equal
+	if weights.count(weights[0]) == len(weights):
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: This might be more idiomatic Python:

`if all(x==weights[0] for x in weights):`

File: fireplace/cards/utils.py [hunk 9]
```diff
@@ -39,52 +41,40 @@
 
 
 # take a list of weights and a list of card pools and produce a random weighted sample without replacement
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: Minor nit: make the comments conform to docstring conventions. Here's a good example:

http://sphinxcontrib-napoleon.readthedocs.org/en/latest/example_google.html

File: fireplace/cards/utils.py (hunk 7 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: totalweight does not have to be an integer.

The other guy used random.random() as well :P

File: fireplace/cards/utils.py [hunk 10]
```diff
@@ -39,52 +41,40 @@
...
-def weighted_card_choice(source, weights, card_sets, count):
+# len(weights) == len(card_sets) (one weight per card set)
+def weighted_card_choice(source, weights: List[int], card_sets: List[str], count: int):
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: Make weights default to `None` and check it in the short-circuit; also make it the last arg

**Comment by smallnamespace** [comment]
Thanks for being patient with me. Would you mind squashing the changes together, and I'll review the package as a whole?

File: fireplace/cards/utils.py (hunk 6 above)
Comment (smallnamespace) [inline fireplace/cards/utils.py]: No, you're right, I'm getting my operators mixed up

**Comment by djkaty** [comment]
Of course

File: fireplace/cards/utils.py (hunk 7 above)
Comment (smallnamespace) [inline fireplace/cards/utils.py]: Yes, I'm just always leery of mixed-mode math because every language seems to have its own quirks.

I'm fine with making the weights floats, but then we should change the type signature

File: fireplace/dsl/random_picker.py [hunk 11]
```diff
@@ -1,79 +1,96 @@
...
 		ret.count = other
 		return ret
 
-	@property
-	def cards(self):
-		if self._cards is None:
-			self._cards = self._filter_cards(self.filters)
-		return self._cards
+	# add a filter set
+	def add(self, weight, **filters):
```
Comment (smallnamespace) [inline fireplace/dsl/random_picker.py]: So the other option is to push the weights straight into the constructor.

I don't have a clear idea of the Discover code you plan to write, but if you will know the weights ahead of time, I think it will be simpler to just push that into __init__ rather than mutating the picker on the fly.

File: fireplace/dsl/random_picker.py [hunk 12]
```diff
@@ -1,79 +1,96 @@
...
-				break
 
 	def __repr__(self):
-		return "%s(%r)" % (self.__class__.__name__, self.filters)
+		return "%s(%r)" % (self.__class__.__name__, self.globalfilter)
 
+	# select number of cards to fetch
 	def __mul__(self, other):
-		ret = self.__class__(*self.args, **self.filters)
+		ret = deepcopy(self)
```
Comment (smallnamespace) [inline fireplace/dsl/random_picker.py]: Curious, was there an issue with shallow copy?

File: fireplace/dsl/random_picker.py [hunk 13]
```diff
@@ -1,79 +1,96 @@
...
+		self.weights.append(weight)
+		self.weightedfilters.append(filters)
+		return self
 
-	def _filter_cards(self, filters):
-		from .. import cards
-		return cards.filter(**filters)
+	# generate a card pool with all cards matching specified (optionally lazy) filters
+	def find_cards(self, source=None, **filters):
+		if len(filters)==0:
```
Comment (smallnamespace) [inline fireplace/dsl/random_picker.py]: `if not filters:`

File: fireplace/dsl/random_picker.py [hunk 14]
```diff
@@ -1,79 +1,96 @@
...
+		# use specific card list if given
+		if cards:
+			self.weights = [1]
+			card_sets = [cards]
 
-	def evaluate(self, source) -> str:
-		if self.lazy_filters:
-			# If the card has lazy filters, we need to evaluate them
-			cards = self.get_cards(source)
+		# use global filters if no weighted filter sets given
```
Comment (smallnamespace) [inline fireplace/dsl/random_picker.py]: Is there a strong reason to distinguish between a global filter and filter sets? If not, I'd advocate just using a singleton filter set with any weight.

File: fireplace/dsl/random_picker.py (hunk 11 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: I don't really like that because of the extra arguments every Random* picker will have unless you specify weights=[,,,] as a keyword argument, but then you'll have to supply all the filters like that too.

The idea is to leave the DSL definitions alone and then in Discover simply take the supplied Random* picker and call eg. ".add(1, card_class=CardClass.NEUTRAL).add(4, card_class=<however we figure out the discover class to use>)". That way it is totally transparent and also flexible for the future.

Comment (djkaty) [inline fireplace/dsl/random_picker.py]: Do bear in mind that __init__ takes a global set of filters and that gets appended with every set of filters added with add(). So for jeweled scarab you call RandomCollectible(cost=3), and this will end up producing two sets of filters:

Weighted 1 - collectible=True, cost=3, card_class=CardClass.NEUTRAL
Weighted 4 - collectible=True, cost=3, card_class=our discover class

So I really don't favour modifying the constructor now.

**Comment by djkaty** [comment]
Immutable weighting function as discussed with JimboHS in IRC.

File: fireplace/dsl/random_picker.py (hunk 12 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: I didn't check but I assumed shallow copy would copy the list pointers/references rather than the actual lists?

File: fireplace/dsl/random_picker.py (hunk 13 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: Does that work if filters exists but is [] ? (again, learning Python...)

File: fireplace/dsl/random_picker.py (hunk 14 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: Discussed in IRC

File: fireplace/dsl/random_picker.py (hunk 13 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: So it does. I've replaced 2 instances of that in the code for the next commit.

File: fireplace/dsl/random_picker.py (hunk 12 above)
Comment (smallnamespace) [inline fireplace/dsl/random_picker.py]: Yes, I think we should avoid mutating the lists within the body of the picker and use `copy` instead. Make copies of the lists within the body of `evaluate` instead.

For example, right now there's no way to re-use the weighted picker because we never reset the weights/lists after a single call to `evaluate`. That's probably not a serious issue, because we don't attempt to evaluate pickers more than once AFAICT, but then you'd need to document properly that evaluate can be called at most once on pickers, and it pushes the onus of copying the picker onto the user, which I'm not a huge fan of.

Comment (djkaty) [inline fireplace/dsl/random_picker.py]: That was an oversight on my part due to lack of knowledge of Python, I agree with this and I've changed it accordingly :)

File: fireplace/cards/brawl/gift_exchange.py [hunk 15]
```diff
@@ -19,8 +19,9 @@ class TB_GiftExchange_Treasure:
 class TB_GiftExchange_Treasure_Spell:
 	# Surely none of this even sort of works.
 	RandomGift = RandomCollectible(
-		COST >= 5,
-		CLASS_CARD=Attr(Controller(OWNER), GameTag.CLASS)
+		card_class=Attr(Controller(OWNER), GameTag.CLASS)
+#		COST >= 5,
+#		CLASS_CARD=Attr(Controller(OWNER), GameTag.CLASS)
```
Comment (jleclanche) [inline fireplace/cards/brawl/gift_exchange.py]: Revert that - I'll take care of it so you can just rebase...

Comment (djkaty) [inline fireplace/cards/brawl/gift_exchange.py]: Okay :) Phew I'm glad you know how to fix it

File: fireplace/dsl/random_picker.py [hunk 16]
```diff
@@ -1,79 +1,97 @@
 import random
 from hearthstone.enums import CardType, Race, Rarity
 from .lazynum import LazyValue
+from copy import copy
```
Comment (jleclanche) [inline fireplace/dsl/random_picker.py]: Mind your import ordering. The copy import should be before the hearthstone import.

File: fireplace/dsl/random_picker.py [hunk 17]
```diff
@@ -1,79 +1,97 @@
...
+from .. import cards
```
Comment (jleclanche) [inline fireplace/dsl/random_picker.py]: This looks unnecessary.

File: fireplace/cards/brawl/gift_exchange.py (hunk 15 above)
Comment (djkaty) [inline fireplace/cards/brawl/gift_exchange.py]: When I commit this revert it will break virtually all of the Travis CI tests just so you know

Comment (jleclanche) [inline fireplace/cards/brawl/gift_exchange.py]: Thats fine.

File: fireplace/dsl/random_picker.py [hunk 18]
```diff
@@ -1,79 +1,97 @@
...
 
 
 class RandomCardPicker(LazyValue):
 	"""
 	Store filters and generate a random card matching the filters on pick()
+	Constructor takes a single global set of filters and uses a default weighting of 1
+	Additional weighted filter sets can be added with add(), these will be merged with the global filters
```
Comment (jleclanche) [inline fireplace/dsl/random_picker.py]: Wrap your docstrings at 82 characters ish to make them more readable.

File: fireplace/dsl/random_picker.py [hunk 19]
```diff
@@ -1,79 +1,97 @@
...
-				break
 
 	def __repr__(self):
-		return "%s(%r)" % (self.__class__.__name__, self.filters)
+		return "%s(%r)" % (self.__class__.__name__, self.globalfilter)
 
+	# select number of cards to fetch
 	def __mul__(self, other):
-		ret = self.__class__(*self.args, **self.filters)
+		ret = copy(self)
```
Comment (jleclanche) [inline fireplace/dsl/random_picker.py]: If you're going to switch to use `copy()`, you should do so in a preliminary commit.

File: fireplace/dsl/random_picker.py [hunk 20]
```diff
@@ -1,79 +1,97 @@
...
-		for k, v in filters.items():
+		for k, v in newFilters.items():
 			if isinstance(v, LazyValue):
-				filters[k] = v.evaluate(source)
-		return self._filter_cards(filters)
+				newFilters[k] = v.evaluate(source)
+
+		return cards.filter(**newFilters)
+
+	# This picks from a single combined card pool without replacement, weighting each filtered set of cards against the total
```
Comment (jleclanche) [inline fireplace/dsl/random_picker.py]: This should be a docstring.

File: fireplace/dsl/random_picker.py [hunk 21]
```diff
@@ -1,79 +1,97 @@
...
+RandomCard = lambda **kw: RandomCardPicker(**kw)
+RandomCollectible = lambda **kw: RandomCardPicker(collectible=True, **kw)
+RandomMinion = lambda **kw: RandomCollectible(type=CardType.MINION, **kw)
+RandomBeast = lambda **kw: RandomMinion(race=Race.BEAST)
+RandomMech = lambda **kw: RandomMinion(race=Race.MECHANICAL)
+RandomMurloc = lambda **kw: RandomMinion(race=Race.MURLOC)
+RandomSpell = lambda **kw: RandomCollectible(type=CardType.SPELL, **kw)
+RandomTotem = lambda **kw: RandomCardPicker(race=Race.TOTEM)
+RandomWeapon = lambda **kw: RandomCollectible(type=CardType.WEAPON, **kw)
+RandomLegendaryMinion = lambda **kw: RandomMinion(rarity=Rarity.LEGENDARY, **kw)
```
Comment (jleclanche) [inline fireplace/dsl/random_picker.py]: Are you 100% certain random pickers will no longer take positional arguments? does this even need to be done?

File: fireplace/dsl/random_picker.py [hunk 22]
```diff
@@ -1,79 +1,97 @@
...
+	def copy_with_weighting(self, weight, **filters):
+		ret = copy(self)
+		ret.weights.append(weight)
+		ret.weightedfilters.append(filters)
+		return ret
 
-	def _filter_cards(self, filters):
-		from .. import cards
-		return cards.filter(**filters)
+	# generate a card pool with all cards matching specified (optionally lazy) filters
```
Comment (jleclanche) [inline fireplace/dsl/random_picker.py]: These types of comments should be docstrings

File: fireplace/dsl/random_picker.py [hunk 23]
```diff
@@ -1,79 +1,97 @@
...
 
+	# select number of cards to fetch
 	def __mul__(self, other):
-		ret = self.__class__(*self.args, **self.filters)
+		ret = copy(self)
 		ret.count = other
 		return ret
 
-	@property
-	def cards(self):
```
Comment (jleclanche) [inline fireplace/dsl/random_picker.py]: Why did you turn the property into `find_cards()`? Having it as a property means a subclass can use a prebuilt list, it was pretty useful.

File: tests/test_dsl.py [hunk 24]
```diff
@@ -29,14 +29,41 @@ def test_empty_selector():
...
+	game.player1.give(WISP)
+
+	spells = len(RandomSpell().find_cards())
+	minions = len(RandomMinion().find_cards())
+	weight = 5
+	ratio = spells / (weight * minions)
+	depth = 100
+	margin = 0.15
+
+	picker = RandomCardPicker().copy_with_weighting(1, type=CardType.SPELL).copy_with_weighting(weight, type=CardType.MINION)
```
Comment (jleclanche) [inline tests/test_dsl.py]: I really don't like the resulting api.

File: fireplace/dsl/random_picker.py (hunk 18 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: Wow, do people still use 640x480? :P Okay, so noted.

File: tests/test_dsl.py [hunk 25]
```diff
@@ -29,14 +29,41 @@ def test_empty_selector():
...
+
+	for i in range(depth):
+		if (picker.evaluate(game.player1.hand[0])[0].type == CardType.SPELL):
+			pickedSpells += 1
+		else:
+			pickedMinions += 1
+
+	pickedRatio = pickedSpells / pickedMinions
+
+	assert abs(pickedRatio - ratio) < margin
```
Comment (jleclanche) [inline tests/test_dsl.py]: These types of tests are slow and inexact. You can leave it out, or you can write a custom script that will do multiple runs and spit out probability tables... the latter would be nice but overkill for this PR.

File: fireplace/dsl/random_picker.py [hunk 26]
```diff
@@ -1,79 +1,97 @@
...
-		self._cards = None
-		self.lazy_filters = False
-		for v in filters.values():
-			if isinstance(v, LazyValue):
-				self.lazy_filters = True
-				break
 
 	def __repr__(self):
-		return "%s(%r)" % (self.__class__.__name__, self.filters)
+		return "%s(%r)" % (self.__class__.__name__, self.globalfilter)
```
Comment (jleclanche) [inline fireplace/dsl/random_picker.py]: Why did you change its name?

File: fireplace/cards/utils.py [hunk 27]
```diff
@@ -38,6 +40,43 @@
...
+		card_set = [j for i in card_sets for j in i]
+		return [source.controller.card(card, source=source) for card in random.sample(card_set, count)]
+
+	chosen_cards = []
+
+	# sum all the weights
+	cum_weights = []
+	totalweight = 0
+	for i, w in enumerate(weights):
+		totalweight += w * len(card_sets[i])
```
Comment (jleclanche) [inline fireplace/cards/utils.py]: You don't need that temporary variable

File: fireplace/cards/utils.py [hunk 28]
```diff
@@ -38,6 +40,43 @@
 AT_MAX_MANA = lambda s: MANA(s) == 10
 
 
+# take a list of weights and a list of card pools and produce a random weighted sample without replacement
+# len(weights) == len(card_sets) (one weight per card set)
+def weighted_card_choice(source, weights: List[int], card_sets: List[str], count: int):
+
+	# special case for single weight (generic random pick)
+	if len(card_sets)==1:
```
Comment (jleclanche) [inline fireplace/cards/utils.py]: Space around operators please

File: fireplace/cards/utils.py [hunk 29]
```diff
@@ -38,6 +40,43 @@
...
+		return [source.controller.card(card, source=source) for card in random.sample(card_sets[0], count)]
```
Comment (jleclanche) [inline fireplace/cards/utils.py]: DRY with the statement below, there's better ways to write that.

File: fireplace/cards/utils.py [hunk 30]
```diff
@@ -38,6 +40,43 @@
 AT_MAX_MANA = lambda s: MANA(s) == 10
 
 
+# take a list of weights and a list of card pools and produce a random weighted sample without replacement
+# len(weights) == len(card_sets) (one weight per card set)
```
Comment (jleclanche) [inline fireplace/cards/utils.py]: docstring

File: tests/test_dsl.py (hunk 24 above)
Comment (djkaty) [inline tests/test_dsl.py]: That's too vague to make a useful response to. The API was discussed for hours with JimboHS last night and this is how he wanted it.

File: fireplace/cards/utils.py (hunk 27 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: Which temporary variable?

File: fireplace/dsl/random_picker.py (hunk 16 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: Fixed in next commit

File: fireplace/dsl/random_picker.py (hunk 17 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: Fixed in next commit

File: fireplace/dsl/random_picker.py (hunk 20 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: Fixed in next commit

File: fireplace/dsl/random_picker.py (hunk 21 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: Discussed in IRC

File: fireplace/dsl/random_picker.py (hunk 22 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: Fixed in next commit

File: tests/test_dsl.py (hunk 25 above)
Comment (djkaty) [inline tests/test_dsl.py]: I agree, I didn't like it either. Luckily I can see from this it does indeed work so I'll take it out.

File: fireplace/dsl/random_picker.py (hunk 26 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: Clarity. There is no longer a single set of filters, there is a main set of filters (globalfilters) and all of the weighted filters to which globalfilters is also applied individually.

File: fireplace/cards/utils.py (hunk 27 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: Discussed in IRC

File: fireplace/cards/utils.py (hunk 28 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: Fixed in next commit

File: fireplace/cards/utils.py (hunk 30 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: Fixed in next commit

File: fireplace/cards/utils.py (hunk 29 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: I've changed the weight line to use set (len(set(weights)) == 1) but the two return statements aren't quite the same, AFAIK any other way of writing it will be more complicated or involve additional assignments or if statements. Please show me a reformulation if there is a Python way I don't know of :)

File: fireplace/dsl/random_picker.py (hunk 23 above)
Comment (djkaty) [inline fireplace/dsl/random_picker.py]: A few reasons:

1. Clarity. There were previously 3 methods that found cards in different ways and via different codepaths. Now there is 1.
2. If a filter was a Lazy filter the cards property would not return the correct values
3. The property is a read-only property anyway
4. find_cards allows you (or a subclass) to specify an optional filter set each time and get a new set of cards back each time. Lazy filters will also work. This can be re-used as much as desired. If no filter is specified, the global filter set is used. You can't do this with a property.
5. Sub-classes can supply their own card lists in the new 'cards' argument to evaluate() if they want to override the list

**Comment by djkaty** [comment]
New commit added. Please re-review!

File: fireplace/cards/utils.py [hunk 31]
```diff
@@ -40,16 +40,19 @@
 AT_MAX_MANA = lambda s: MANA(s) == 10
 
 
-# take a list of weights and a list of card pools and produce a random weighted sample without replacement
-# len(weights) == len(card_sets) (one weight per card set)
 def weighted_card_choice(source, weights: List[int], card_sets: List[str], count: int):
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: Did we decide that the weights could be floats? If so, `List[float]`

File: fireplace/cards/utils.py [hunk 32]
```diff
@@ -38,6 +40,46 @@
...
+	sample without replacement.
+	len(weights) == len(card_sets) (one weight per card set)
+	"""
+
+	# special case for single weight (generic random pick)
+	if len(card_sets) == 1:
+		return [source.controller.card(card, source=source) for card in random.sample(card_sets[0], count)]
+
+	# special case for all weights equal
+	if len(set(weights)) == 1:
```
Comment (smallnamespace) [inline fireplace/cards/utils.py]: This case smells like too much of premature optimization to me.

We're only doing a small bit of extra book-keeping in the general case. Would you be willing to micro-bench this function with and without this shortcut to see how much of a difference it makes?

File: fireplace/cards/utils.py (hunk 31 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: Weights are floats, so noted.

File: fireplace/cards/utils.py (hunk 32 above)
Comment (djkaty) [inline fireplace/cards/utils.py]: Either you or jleclanche asked me to put this in, so...

Comment (smallnamespace) [inline fireplace/cards/utils.py]: I think the special case for a single weight is fine, but not this one.

In any case:

1. There's good reason to expect that both code paths have almost the same runtime
2. This is emphatically *not* in an inner loop of fireplace

If you really care about performance, profile or benchmark and then optimize where it will actually matters. Otherwise you'll end up wasting a lot of effort in optimizing things that *don't matter at all to actual performance*.

Comment (jleclanche) [inline fireplace/cards/utils.py]: I didn't ask you to put this in; you said you *wanted* to put this in and I didn't say no. I agree with @smallnamespace and think this is premature optimization.

Comment (djkaty) [inline fireplace/cards/utils.py]: Just read this. Do you want the 2nd if condition removed or both of them?

**Comment by djkaty** [comment]
https://github.com/jleclanche/fireplace/pull/346

**Comment by djkaty** [comment]
It should be good now, yay!
